[wasm-ep] Minimal diagnostic tracing configuration and sample

lambdageek opened the pull request:
* Adds a `/p:WasmEnablePerfTracing=true` configuration.

   In this configuration the runtime is built with threading (`MonoWasmThreads` property is true), but C# code is not allowed to start threads and doesn't use the portable thradpool (`MonoWasmThreadsNoUser` property is also true).

   The upshot is that EventPipe can start threads but user code is still single-threaded.

* Adds a `MONO.diagnostics` interface in JavaScript.  There's a single method for now `createEventPipeSession` which creates a session that can save a trace to a file on the virtual file system.  JS code (or a user in a dev console) needs to call `start()` and `stop()` on the session object to begin collecting samples.    For now there's no way to specify any configuration options other than the destination VFS file path for the trace file.

* Adds a sample that runs an async task for five seconds and collects samples and then triggers a `click()` to download the trace file out of the browser.

lambdageek committed:
add DISABLE_WASM_USER_THREADS mono cmake option

lambdageek committed:
Disable Thread.StartInternal icall if DISABLE_WASM_USER_THREADS

if threading is enabled for the runtime internally, but disabled for
user code, throw PNSE

lambdageek committed:
add an eventpipe sample

lambdageek committed:
[wasm-ep] (browser-eventpipe sample) run loop for longer

lambdageek committed:
[samples/wasm-eventpipe] make an async task sample

change the sample to do some work asynchronously using setTimeout
instead of blocking

lambdageek committed:
[wasm] Add MONO.diagnostics interface

Binds enable, start, disable methods defaulting to non-streaming FILE mode

lambdageek committed:
if wasm threads are disabled, but perftracing is enabled, don't log
overlapped io events

lambdageek committed:
fix whitespace and nits

lambdageek committed:
don't need try/finally in the sample anymore

lambdageek commented:
/cc @kg @lateralusX

lambdageek committed:
more whitespace

lambdageek committed:
add start method to EventPipeSession interface

lambdageek commented:
/cc @pavelsavara

pavelsavara commented on src/mono/wasm/runtime/types.ts:
```diff
@@ -173,6 +173,12 @@ export type CoverageProfilerOptions = {
 
+/// Options to configure the event pipe session
+export interface EventPipeSessionOptions {
+    /// location on the VFS where the session trace will be saved
```
VFS is good for us, but most developers would not know how to access it. Could we optionally lift the data from VFS and create download from it via [`<a href="data:application/octet-stream;` ](https://stackoverflow.com/a/18197341/129124)
On the API it could be `downloadTraceAfterSessionEnd:boolean` and work only in the browser. In node it could save directly to real disk.

it's just idea, not blocking the merge

lambdageek replied:
Yea I do something like that in the sample.  Easy enough to promote to an API

lambdageek replied:
I ended up going another way.  Added `getTraceBlob()` and `getTraceDataURI()` methods to the session object.

tracing and saving can be done like this:

```javascript
   const session = MONO.diagnostics.createEventPipeSession();
   session.start ();
   // do work
   session.stop ();
   const blob = session.getTraceBlob();
   const url = URL.createObjectURL (blob);
   downloadLink (url, "local-filename.nettrace");  // makes an <a download="..." href="object:..." /> and clicks on it
   URL.revokeObjectURL (url);
```

To be honest I'm not sure if the data URI one is needed.  This stuff depends on SharedArrayBuffer and web workers, so it needs newish browsers anyway.

lambdageek committed:
don't run wasm-eventpipe sample on CI lanes without perftracing

lambdageek committed:
more whitespace

lambdageek committed:
fix eslint warnings, default rundown to true, allow callback for traceFilePath option

lambdageek committed:
add EventPipeSession.getTraceBlob and EventPipeSession.getTraceDataURI

two ways of retreiving the collected traces instead of exposing the
emscripten VFS directly.

Probably the Blob one is enough.  Is there any reason to also provide
a data URI?

update the sample to use URL.createObjectURL (session.getTraceBlob())
to create the download link

lateralusX commented on src/mono/mono/component/event_pipe-stub.c:
```diff
@@ -6,6 +6,35 @@
+			     /* IpcStream stream = NULL, */
+			     /* EventPipeSessionSycnhronousCallback sync_callback = NULL, */
+			     /* void *callback_additional_data, */
+			     int32_t *out_session_id);
```
Only support for 32-bit session id's? EnvetPipe uses 64-bit id's and it is mapped to/from a pointer to underlying session struct.

lambdageek replied:
did I get it wrong? I thought the session ID was just a pointer cast to an int.  (wasm is 32-bit).

but yea we can use a wider session ID

lateralusX replied:
Under the hood its a pointer casted to an EventPipeSessionID that is typed to uint64_t on all platforms.

lateralusX commented on src/mono/mono/component/event_pipe-stub.c:
```diff
@@ -6,6 +6,35 @@
+G_BEGIN_DECLS
+
+EMSCRIPTEN_KEEPALIVE gboolean
+mono_wasm_event_pipe_enable (const char *output_path,
```
nit: event_pipe-stub.c uses ep types so const ep_char8_t

lateralusX reviewed (commented):
The EventPipe bits looks good, just a couple of small comments/thoughts.

lambdageek committed:
[browser-eventpipe sample] remove unnecessary ref assemblies

lambdageek committed:
use toBase64StringImpl instead of btoa

lambdageek committed:
use int64_t for event pipe session IDs

lambdageek committed:
use ep_char8_t for C decls of event pipe wasm exports

lambdageek committed:
fix debug output

lambdageek committed:
fix endianness issue

lambdageek committed:
Use stack allocation for temporaries

Expose the emscripten stack allocation API

lambdageek commented:
@kg @pavelsavara  I added some stuff that will have broader impact:
1. A new `cuint64.ts` module that stores it as two 32-bit pieces.  Also added `getCUInt64` and `setCUIn64` to `memory.ts` to read and write 64-bit ints out of the emscripten heap.  There's some conversions to `number` (dangerous)` and `bigint` although I don't make use of these except for debugging.
2. Exposed `stackAlloc`, `stackSave`, `stackRestore` in `emscripten.ts`.  I use these to temporarily stack allocate an int64_t to pass the event pipe session ID between JS and native.

pavelsavara commented:
> 1. A new `cuint64.ts` module that stores it as two 32-bit pieces.  
I had this PR some time ago and will need to finish it soon and it's in the same area.
https://github.com/dotnet/runtime/pull/64845

I need to be able to convert C# `long` into JS `number` as part of the new marshaling.
But I need it to use all 52 integral bits of the underlying double. `Number.MAX_SAFE_INTEGER`
Beyond that range, it should round to next possible integer, and not go crazy by setting exponent bits.

My naïve implementation uses BigInt to do the conversion, but it allocates object. 
So I want to improve and use math+shifts instead.

Your `CUInt64` is also expensive by allocating array.  
I wonder if we could use 'Number' and it's 64 bits to store the bits and only do some math when we need full sized BigInt.
But that would be different math than in my use case, because we would store different bits.
And different math to get `Number`.

Thoughts ?

> 2. Exposed `stackAlloc`, `stackSave`, `stackRestore` 
That's OK. Soon I will need them too.

lambdageek commented:
> > 1. A new `cuint64.ts` module that stores it as two 32-bit pieces.
> 
> I had this PR some time ago and will need to finish it soon and it's in the same area. #64845
> 
> I need to be able to convert C# `long` into JS `number` as part of the new marshaling. But I need it to use all 52 integral bits of the underlying double. `Number.MAX_SAFE_INTEGER` Beyond that range, it should round to next possible integer, and not go crazy by setting exponent bits.
> 
> My naïve implementation uses BigInt to do the conversion, but it allocates object. So I want to improve and use math+shifts instead.
> 
> Your `CUInt64` is also expensive by allocating array. 

> I wonder if we could use 'Number' and it's 64 bits to store the bits and only do some math when we need full sized BigInt.

I guess the idea is to do something like this:
```typescript

/// We pack 64-bit integers into the 64 bytes of a double
type PackedInt64 = number;

const doubleView = new Float64Array(1);
const int32View = new Int32Array(doubleView.buffer);

export function pack(lo: number, hi: number): PackedInt64 {
    int32View[0] = lo;
    int32View[1] = hi;
    return doubleView[0];
}

exort function getLo(value: PackedInt64): number {
    doubleView[0] = value;
    return int32View[0];
}

export function getHi(value: PackedInt64): number {
    doubleView[0] = value;
    return int32View[1];
}

export function toBigInt(value: PackedInt64): bigint {
    return BigInt(getLo(value)) + BigInt(getHi(value)) << BigInt(32);
}
```

And this works on Chrome:
```console
> let x = pack (12, 0xffffffff)
undefined
> x
NaN
> getLo(x)
12
> getHi(x)
-1
```

But not in Firefox because of [NaN boxing](https://brionv.com/log/2018/05/17/javascript-engine-internals-nan-boxing/):

```console
> let x = pack(12,0xffffffff)
undefined
> x
NaN
> getLo(x)
0
> getHi(x)
2146959360
​```

because they rewrite all NaNs into a canonical form, so `y = pack (1234, 0xffffffff)` will also have `getLo(y) == 0` and `getHi(y) == 2146959360`

> But that would be different math than in my use case, because we would store different bits. And different math to get `Number`.

I'm not sure if it's possible just using math - for int64 values that have a bit pattern of a NaN, I'm not sure how you would construct it just using math.

> Thoughts ?

Ask google to optimize the heck out of BigInt for small (<=64 bit) values

pavelsavara commented:
the `toBigInt` could be implemented by `BigInt64Array` view on your `buffer`.

On firefox and `NaN`, do they normalize on the write to the `Float64Array` buffer or on the pulling the `Number` out of the buffer or both ? Perhaps making Float64Array a view and Int32Array the buffer could help ?

lambdageek commented:
> the `toBigInt` could be implemented by `BigInt64Array` view on your `buffer`.
> 
> On firefox and `NaN`, do they normalize on the write to the `Float64Array` buffer or on the pulling the `Number` out of the buffer or both ? Perhaps making Float64Array a view and Int32Array the buffer could help ?

as far as I can tell, it gets normalized when pulling the double out of the Float64Array.  The buffer contents don't change (ie: the int32View still gets the same bits that were stored in there).  Changing whether the Int32Array or the Float64Array creates the underlying buffer doesn't make a difference.

Doing something like `doubleView[0] = doubleView[0]` will change the bits in the buffer.  (I believe because the RHS `doubleView[0]` is where they do the renormalization when they create the temporary `number`)

lambdageek committed:
Use 32-bit EventPipe session ID on WASM

64 bit integers are awkward to work with in JavaScript.

The EventPipe session ID is derived from a pointer address, so even
though it is nominally a 64-bit value, in practice the top bits are
zero.

Use a 32-bit int to represent the session ID on the javascript side
and convert to 64-bit in C when calling down to the EventPipe APIs

lambdageek committed:
Make the sample do more work in managed

give the sample profiler some non-empty samples to collect

lambdageek commented:
I left `cuint64.ts` in the PR (I can remove it if we prefer that), but it's not used for EventPipe anymore.

EventPipe session IDs are stored as 64-bits in the runtime, but the actual values are derived from a pointer address, so the top 32-bits are zero (and I assert the same when I start a session).  Instead of doing awkward 64-bit interop, I now just use a 32-bit number for the session ID on the JS side, and pad it out to 64-bits on the C side before calling the EventPipe APIs.

lambdageek commented:
/azp run runtime-wasm

kg commented on src/mono/cmake/options.cmake:
update the message to mention that this only applies to wasm

lambdageek replied:
Updated.  also clarified that runtime internal managed threads will be allowed (I need them to support the EventPipe counter sampling thread).

pavelsavara commented on src/mono/sample/wasm/browser-eventpipe/main.js:
```diff
@@ -0,0 +1,101 @@
+    wasm_exit(exit_code);
+}
+
+setTimeout(main, 10000);
```
This confuses me, wait 10 seconds and then run `main` ? Why ?

pavelsavara replied:
`console.log("waiting for curious human before starting the program")`

pavelsavara commented on src/mono/wasm/runtime/diagnostics.ts:
```diff
@@ -0,0 +1,142 @@
+    Done,
+}
+
+function withStackAlloc<TRes> (bytesWanted: number, f : (ptr: VoidPtr) => TRes): TRes {
```
Nice, please move it to `memory.ts` 
@kg do you think that allocating the callback closure is too expensive for general usage ?

kg replied:
yes, but it doesn't seem to be used in hot paths so this is probably fine

kg replied:
https://github.com/kg/runtime/commit/2c62ddf2c541f515714532afd53bc7b944a871eb#diff-d320e7f4617dada664d0674a4bd8ee2a1e8eae9118f7b4320112e70d20ebfb49R106-R109 is an example of one way to avoid the closure by adding userdata arguments instead

lambdageek replied:
ok, I added declarations for up to 7 UD arguments.  

But it's kind of hard to avoid creating a closure in TS/JS because even module imports are closures.

lambdageek replied:
makes sense, thank you. I rewrote my use of `withStackAlloc` to use a helper function for the callback. (I don't expect EP sessions to be created that aggressively, but I want to get the muscle memory for doing things right for other stuff that might be on the hot path)

lambdageek commented on src/mono/mono/component/event_pipe-stub.c:
```diff
@@ -4,6 +4,7 @@
 
 #include <config.h>
 #include "mono/component/event_pipe.h"
+#include "mono/component/event_pipe-wasm.h"
```
The header itself has the ifdefs. Could do it either way, but this way the ifdefs are in one place instead of in every .c file. but I'm not opposed to changing it

pavelsavara commented on src/mono/wasm/runtime/diagnostics.ts:
```diff
@@ -0,0 +1,142 @@
+}
+
+// The name of the session trace file on the VFS.  This is an implementation detail that is not exposed to the user.
+function computeTracePath (): string {
```
nit: could be inlined.

lambdageek replied:
I decided to just use consecutive numbering for the VFS files.  the timestamp wasn't effective anyway if sessions are created below the timestamp resolution.

lambdageek committed:
Move withStackAlloc to memory.ts

pavelsavara commented:
ts part LGTM.

follow-up could be adding an method for javaScript to contribute an event ?
For example, interop related metrics.
I'm not sure that should be public/embedding API.

lambdageek committed:
fix option description

lambdageek committed:
simplify VFS .nettrace file naming

Just use consecutive integers to uniquify the session traces.  Dont'
need a fancy timestamp in the VFS (which would also not be unique if
you create sessions below the timestamp resolution)

lambdageek committed:
Add overloads to memory.withStackAlloc to avoid creating closures

Pass up to 7 additional arguments to the callback function

lambdageek committed:
sample: explain why there's a 10s pause

lambdageek committed:
move createEventPipeSession callback to a function

ensures the closure is created once

lambdageek commented:
/azp run runtime-wasm

lambdageek commented:
Any other feedback?  
/cc @lewing

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-ep

pavelsavara commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -114,3 +115,37 @@ export function getF32(offset: _MemOffset): number {
+export function withStackAlloc<T1,T2,T3,T4,TResult> (bytesWanted: number, f: (ptr: VoidPtr, ud1: T1, ud2: T2, ud3: T3, ud4: T4) => TResult, ud1: T1, ud2: T2, ud3: T3, ud4: T4): TResult;
+export function withStackAlloc<T1,T2,T3,T4,T5,TResult> (bytesWanted: number, f: (ptr: VoidPtr, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5) => TResult, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5): TResult;
+export function withStackAlloc<T1,T2,T3,T4,T5,T6,TResult> (bytesWanted: number, f: (ptr: VoidPtr, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5, ud6: T6) => TResult, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5, ud6: T6): TResult;
+export function withStackAlloc<T1,T2,T3,T4,T5,T6,T7,TResult> (bytesWanted: number, f: (ptr: VoidPtr, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5, ud6: T6, ud7:T7) => TResult, ud1: T1, ud2: T2, ud3: T3, ud4: T4, ud5: T5, ud6: T6, ud7: T7): TResult;
```
Also, isn't that last one duplicate ?

lambdageek replied:
In the implementation type the user args are `udN ?: TN` in the specialization types they're not `undefined`.

But actually, we can just use tuples:

```typescript
export function withStackAlloc<TArgs extends any[], TResult> (bytesWanted: number, f : (ptr: VoidPtr, ...userargs : [...TArgs]) => TResult, ... userargs : [... TArgs]): TResult {
    const sp = Module.stackSave();
    const ptr = Module.stackAlloc (bytesWanted);
    try {
        return f (ptr, ... userargs);
    } finally {
        Module.stackRestore(sp);
    }
}
```

Unless there's some reason not to use rest args, I would like to commit the above version instead.  @pavelsavara @kg

lambdageek committed:
Merge branch 'main' into wasm-ep

lambdageek committed:
fix whitespace

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
Use a tuple type for withStackAlloc

lambdageek committed:
fix whitespace

lambdageek committed:
go back to explicit args for withStackAlloc insead of rest/spread

kg commented on src/mono/mono/component/event_pipe.c:
```diff
@@ -327,3 +330,73 @@ mono_component_event_pipe_init (void)
+ep_to_wasm_session_id (EventPipeSessionID session_id)
+{
+	g_assert (0 == (uint64_t)session_id >> 32);
+	return (uint32_t)session_id;
```
Why is this cast necessary? EventPipeSessionId is uint32_t

kg replied:
Since the return type is MonoWasmEventPipeSessionID maybe it should be a cast to that. Also, if EventPipeSessionID becomes 64-bit later and the return value is somehow 32-bit, I think this would be a wrapping conversion which could cause a non-zero EventPipeSessionID to become 0? I guess that's what the assert is for but it feels gross to do the assert on something other than the actual return value, I would prefer to assert on the (post-cast) return value instead.

lambdageek replied:
`EventPipeSessionID` is `uint64_t`; `MonoWasmEventPipeSessionID` is `uint32_t`.  The first cast that is redundant, but I thought it'd be good to have it there for documentation.

kg commented on src/mono/mono/metadata/threads.c:
```diff
@@ -4815,7 +4815,7 @@ ves_icall_System_Threading_Thread_StartInternal (MonoThreadObjectHandle thread_h
 
-#ifdef DISABLE_THREADS
+#if defined (DISABLE_THREADS) || defined (DISABLE_WASM_USER_THREADS)
 	mono_error_set_platform_not_supported (error, "Cannot start threads on this runtime.");
```
We should think about updating this message since it is now specifically about user/managed threads. Not important to do right away though

kg commented on src/mono/sample/wasm/browser-eventpipe/Program.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        public static void Main(string[] args)
+        {
+            Console.WriteLine ("Hello, World!");
```
Is it intended that nothing actually happens when you run this program?

kg replied:
As mentioned in the main.js review, maybe put a single-line comment here to indicate that the real stuff happens in main.js

lambdageek replied:
This is some kind of weird Wasm sample build situation: Roslyn needs me to have a `Main` because I don't have `OutputType` specified in the .csproj.  but if I set `OutputType` to `Library` and make Roslyn happy, publishing gets upset and the app doesn't build.

But `Main` never actually gets called.

kg commented on src/mono/wasm/runtime/diagnostics.ts:
```diff
@@ -0,0 +1,128 @@
+    const rundown = options?.collectRundownEvents ?? defaultRundownRequested;
+
+    memory.setI32(sessionIdOutPtr, 0);
+    if (!cwraps.mono_wasm_event_pipe_enable(tracePath, defaultBufferSizeInMB, defaultProviders, rundown, sessionIdOutPtr)) {
```
Is it intended that the defaults are used instead of being read from the options?

lambdageek replied:
yea I didn't bind all the options yet into EventPipeSessionOptions.  The next PR will expand the options type providers will come from there.  I'm not sure yet what effect "buffer size in MB" has, so that may or may not get exposed too.

lambdageek commented on src/mono/wasm/runtime/cuint64.ts:
```diff
@@ -0,0 +1,48 @@
+/// and 'import type { CUInt64 } from './cuint64';
+export type CUInt64 = readonly [number, number];
+
+export function toBigInt (x: CUInt64): bigint {
```
We discussed it. Once there is faithful marshaling of `uint64_t`, this is less necessary.  but for now for situations where we control the signature on the C side it's one way to work with a `uint64_t *` argument and get all the bits back and forth.

kg commented on src/mono/wasm/runtime/types/emscripten.ts:
```diff
@@ -62,4 +66,4 @@ export declare interface EmscriptenModule {
 
-export declare type TypedArray = Int8Array | Uint8Array | Uint8ClampedArray | Int16Array | Uint16Array | Int32Array | Uint32Array | Float32Array | Float64Array;
\ No newline at end of file
+export declare type TypedArray = Int8Array | Uint8Array | Uint8ClampedArray | Int16Array | Uint16Array | Int32Array | Uint32Array | Float32Array | Float64Array;
```
You can also use ArrayBufferView in place of this, though it has the downside of that interface inexplicably not having .length. Just an FYI, I'm fine with us defining a TypedArray alias (do the [bigint arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt64Array) need to go in here too?)

lambdageek committed:
use unsigned 32-bit get/set in cuint64 get/set

lambdageek committed:
cosmetic changes to sample

lewing reviewed (approved):
:shipit:
